clang/labs/lab1/kosarev_egor/always_inline.cpp

```diff
@@ -0,0 +1,70 @@
+#include "clang/AST/ASTConsumer.h"
+#include "clang/AST/RecursiveASTVisitor.h"
+#include "clang/Frontend/CompilerInstance.h"
+#include "clang/Frontend/FrontendPluginRegistry.h"
+
+class AddAlwaysInlineConsumer : public clang::ASTConsumer {
+public:
+  bool hasCondition(clang::Stmt *S) {
+    if (!S) {
+      return false;
+    }
+
+    if (llvm::isa<clang::IfStmt>(S) || llvm::isa<clang::SwitchStmt>(S) ||
+        llvm::isa<clang::WhileStmt>(S) || llvm::isa<clang::DoStmt>(S) ||
+        llvm::isa<clang::ForStmt>(S)) {
+      return true;
+    }
+
+    for (clang::Stmt *Child : S->children()) {
+      if (hasCondition(Child)) {
+        return true;
+      }
+    }
+
+    return false;
+  }
+
+  bool HandleTopLevelDecl(clang::DeclGroupRef D) override {
+    clang::FunctionDecl *FD = nullptr;
+
+    for (clang::Decl *Decl : D) {
+      FD = clang::dyn_cast<clang::FunctionDecl>(Decl);
+      if (FD) {
+        if (!FD->hasAttr<clang::AlwaysInlineAttr>() &&
+            !hasCondition(FD->getBody())) {
+          FD->addAttr(
+              clang::AlwaysInlineAttr::CreateImplicit(FD->getASTContext()));
+          FD->print(llvm::outs() << "__attribute__((always_inline)) ");
+        } else
+          FD->print(llvm::outs());
+      }
+    }
+    return true;
+  }
+};
+
+class AddAlwaysInlineAction : public clang::PluginASTAction {
+protected:
+  std::unique_ptr<clang::ASTConsumer>
+  CreateASTConsumer(clang::CompilerInstance &Compiler,
+                    llvm::StringRef InFile) override {
+    return std::make_unique<AddAlwaysInlineConsumer>();
+  }
+
+  bool ParseArgs(const clang::CompilerInstance &CI,
+                 const std::vector<std::string> &Args) override {
+    for (const std::string &arg : Args) {
+      if (arg == "--help") {
+        llvm::outs() << "This plugin adds the always_inline attribute to "
+                        "functions if they do not have conditions!\n";
+      } else
+        llvm::outs()
+            << "Use the --help argument to understand the plugin's purpose!\n";

```

Comment:
Why it returns true if `help` or unexpected attribute is specified?

Reply:
The program returns true if the "--help" attribute or other unexpected attribute is specified, because inside the `ParseArgs` method, all arguments except "--help" are processed so that the message "Use the --help argument to understand the plugin's purpose!" is displayed and it returned true, indicating that the argument processing was successful.

Reply:
But if argument is incorrect or it is just `help` do we still execute plugin action?

Reply:
If the argument is incorrect or it is just "–help", then the plugin action will not be executed. In the ParseArgs method, all arguments except "–help" are processed to display the message "Use the –help argument to understand the plugin's purpose!". If the argument does not match what is expected or is just "–help", then the plugin action will not be performed, and the program will return false, indicating that the processing of arguments failed.

Reply:
I don't see a line which stops plugin execution if 'help' is specified.

Reply:
I incorrectly implemented the processing, redid it